File: 0525-contiguous-array/0525-contiguous-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function(nums) {
    let res = 0;
    let zero = 0;
    let one = 0;
    let map = new Map();
    map.set(0, 1);

    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === 0) {
            zero++;
        } else {
            one++;
        }

        let diff = zero - one;
        if (diff === 0) {
            res = Math.max(res, i + 1);
            continue;
        }

        if (map.has(diff)) {
            let idx = map.get(diff);
            let len = i - idx;
            res = Math.max(res, len);
        } else {
            map.set(diff, i);
        }
    }
    return res;
};

File: 0525-contiguous-array/0525-contiguous-array.go
func findMaxLength(nums []int) int {
    zero := 0
    one := 0
    res := 0
    mp := make(map[int]int)
    mp[0] = 1

    for i := 0; i < len(nums); i++ {
        if nums[i] == 0 {
            zero++
        } else {
            one++
        }

        diff := zero - one
        if diff == 0 {
            res = max(res, i + 1)
            continue
        }

        if idx, exist := mp[diff]; exist {
            res = max(res, i - idx)
        } else {
            mp[diff] = i
        }
    }

    return res
}